packages/aws-cdk/lib/api/hotswap/monitor/logs-monitor.ts

```diff
@@ -0,0 +1,226 @@
+import * as util from 'util';
+import * as colors from 'colors/safe';
+import { error } from '../../../logging';
+import { ISDK } from '../../aws-auth';
+
+export interface WithDefaultPrinterProps {
+  /**
+   * Time that the latest hotswap operation occurred
+   *
+   * @default - local machine's current time
+   */
+  readonly hotswapTime?: Date;
+
+}
+
+export interface CloudWatchLogEvent {
+  readonly message: string;
+  readonly logGroup: string;
+  readonly logStream?: string;
+  readonly ingestionTime: Date;
+}
+
+export class CloudWatchLogEventMonitor {
+  public static withDefaultPrinter(options: WithDefaultPrinterProps = {}) {
+    const stream = process.stderr;
+    const props: PrinterProps = {
+      stream,
+    };
+
+    return new CloudWatchLogEventMonitor(new EventPrinter(props), options.hotswapTime);
+  }
+
+  private active = false;
+
+  /**
+   * Determines which events not to display
+   */
+  private startTime: number;
+
+  /**
+   * Current tick timer
+   */
+  private tickTimer?: NodeJS.Timer;
+
+  /**
+   * Set to the activity of reading the current events
+   */
+  private readPromise?: Promise<any>;
+
+  private activity = new Set<string>();
+
+  private logGroups = new Set<string>();
+  private sdk?: ISDK;
+
+  constructor(
+    private readonly printer: IEventPrinter,
+    hotswapTime?: Date,
+  ) {
+    this.startTime = hotswapTime?.getTime() ?? Date.now();
+  }
+
+  public start() {
+    this.active = true;
+    this.scheduleNextTick();
+    return this;
+  }
+
+  public addSDK(sdk: ISDK) {
+    this.sdk = sdk;
+  }
+
+  /**
+   * Adds a CloudWatch log group to read log events from
+   */
+  public addLogGroups(logGroups: string[]) {
+    logGroups.forEach(group => {
+      if (!this.logGroups.has(group)) {
+        this.logGroups.add(group);
+      }
+    });
+  }
+
+  public async stop() {
+    this.active = false;
+    if (this.tickTimer) {
+      clearTimeout(this.tickTimer);
+    }
+  }
+
+  private scheduleNextTick() {
+    if (!this.active) {
+      return;
+    }
+    this.tickTimer = setTimeout(() => void (this.tick()), this.printer.updateSleep);
+  }
+
+  private async tick() {
+    if (!this.active) {
+      return;
+    }
+    try {
+      this.readPromise = this.readNewEvents();
+      await this.readPromise;
+      this.readPromise = undefined;
```

Comment:
I have to say, this code is a little baffling to me 😛. Why not just:

```suggestion
      const events = await this.readNewEvents();
```